### ISSUE TITLE
minimega: refactor all vm info fields (and any other things we can find) into the VM's vmInfo struct

### ISSUE THREAD

**Issue by djfritz** [sandia-minimega/minimega#165]
We have many correlated data structures as globals, and many of them are just linked to a particular VM. Eliminate the dispersed data and move vm specifics into the vmInfo struct.

IF WE CAN. There may be some concerns by doing this. Some things wait on vms to die/quit, and gracefully shut other things down. We need to be careful about holding on to vmInfo handles in that case.

**Comment by floren** [comment]
Any update on this? If you have some specific data structures in mind I might take a crack at it too.

**Comment by djfritz** [comment]
There may be a larger refactor looming if we decide to add containers in the way I'm thinking. Let's hold off on this until we scope the containers stuff. Probably a few weeks on this still...

**Comment by djfritz** [comment]
Other big things on the horizon may obviate this, as a bigger refactor may be due. Pushing to 2.1

**Comment by jcrussell** [comment]
IP/IPv6 shouldn't be too hard to fix. Proposal:

Create channel that sends (mac, *IP) pairs from ipmac back to the bridge code. The bridge code can then find the appropriate VM and update the (to-be-added) IP field of NetConfig.

**Comment by jcrussell** [comment]
IP4/IP6 should be ready to go (b332a37e81992283bed38d7fb064fe8593ed210f).

The bandwidth stats look like they will introduce quite a bit more overhead since they are updated every five seconds. @djfritz: any ideas on how to cleanly implement this?

**Comment by djfritz** [comment]
This is a tough one. We used to reload bandwidth stats on demand, but the API call can become expensive. We moved it out of band with the 5 second refresh, and expanded it somewhat to generate per interface statistics. That's nice to have, but obviously won't scale well when we get things like containers in.

So there are 3 solutions I see:

1. Stop gathering bw statistics in this fashion altogether (though I see no better solution).
2. Increase the collection interval to 15-30 seconds?
3. Keep this statistic out of the vm struct, instead adding a GetBW() method to the vm interface with a strong comment detailing the expense of bw accounting and why it's not stored with the VM. 

Maybe the solution is a combination of 2 and 3? Thoughts?

**Comment by jcrussell** [comment]
Perhaps we're thinking about this wrong, should the bandwidth stats be included in `vm info`? My understanding is that if you create one tap/vm then the bandwidth stats will be per vm. Do all of our users do that? Maybe it would be better to expand the existing `host bandwidth` output to include per tap stats and then an enterprising user can do the join.

Alternatively, we could make the collection interval more dynamic by recording when we last read the tx/rx bytes for each interface and only sleeping we go to read the tx/rx bytes again and the interval to the previous reading is shorter than our minimum collection interval. This way, if updates take a little longer to update all the vms when there are more vms, we will still do accurate accounting.

**Comment by djfritz** [comment]
Maybe meet halfway on your first point - we already gather per tap stats, so maybe the best organization here is simply:

* bandwidth code gathers statistics and stores per tap data in its own structures
* when updating, do a reverse lookup just to note which vm owns the tap
* on printing information, group taps by like vm

**Comment by jcrussell** [comment]
So does that mean that we don't include the bandwidth stats in `vm info`?

Instead we would have `tap bandwidth` output something like:

```
tap  | tx            | rx            | vms 
tap0 | xxx (xx MB/s) | yyy (yy MB/s) | foo[2-4],bar
tap0 | uuu (uu MB/s) | vvv (vv MB/s) | foo4,car
```

Where a VM can be listed under multiple taps? We could also split the tx/rx bytes from the current rate.

**Comment by djfritz** [comment]
vms cannot share taps - a tap for kvm is a pipe with a specific qemu holding one end of it, so we don't have to worry about that.

grumble grumble - let's stew on this one a bit. I'd really like to have the bw stats in vm info - it would make monitoring a vm with some external tool simpler.

**Comment by jcrussell** [comment]
In an offline conversation, we decided to pull this column out of VM info. We will add the VM ID and name to `tap bandwidth`.

**Comment by jcrussell** [comment]
Marking this as closed with the merge of #229.
